### PR TITLE
Flock floor/wall running improvements

### PR DESCRIPTION
**Pull request by amylizzle**
## About the PR 
Enables the AI to path through converted walls.
Makes human controlled drones pass through walls in a better way (ie, so long as a wall is converted, hold shift to run through it)

## Why's this needed? 
More flock buffs, less unintuitive behaviour

### CODE REVIEW

File: _std/pathfinding.dm [hunk 1]
```diff
@@ -413,6 +413,11 @@
 /// Returns false if there is a dense atom on the turf, unless a custom hueristic is passed.
 /proc/jpsTurfPassable(turf/T, turf/source=null, atom/passer=null, id=null)
 	. = TRUE
+	if(istype(passer,/mob/living/critter/flock/drone) && istype(T, /turf/simulated/wall/auto/feather))
+		var/mob/living/critter/flock/drone/F = passer
+		if(F.floorrunning || (F.can_floorrun && F.resources >= 1))
+			return TRUE // floor running drones can *always* pass through flockwalls
```
Comment (FlameArrow57) [inline _std/pathfinding.dm:419]: Also needs to make sure the wall isn't broken?

File: code/turf/turf_flock.dm [hunk 2]
```diff
@@ -356,6 +356,19 @@
 		else if(!isfeathertile(newloc))
 			F.end_floorrunning()
 
+/turf/simulated/wall/auto/feather/Bumped(AM)
+	. = ..()
+	if(istype(AM, /mob/living/critter/flock/drone))
+		var/mob/living/critter/flock/drone/F = AM
+		if(F.floorrunning || (F.can_floorrun && F.resources >= 1))
```
Comment (FlameArrow57) [inline code/turf/turf_flock.dm:363]:
```suggestion
		if((F.floorrunning || F.can_floorrun) && F.resources >= 1)
```
Shouldn't be able to floorrun with < 1 resource, but otherwise seems like saying you can?

File: code/turf/turf_flock.dm [hunk 3]
```diff
@@ -356,6 +356,19 @@
...
+			if(!F.client || (F.client && F.client.check_key(KEY_RUN))) //ai doesn't have to hold shift to wallrun, people do
+				F.start_floorrunning()
+				F.set_loc(src)
```
Comment (FlameArrow57) [inline code/turf/turf_flock.dm:366]:
```suggestion
				F.set_loc(src)
				F.start_floorrunning()
```
Not sure if it matters, but feels bit off saying floorrun on turf that may not be floorrunnable/not the turf you are floorrunning on

File: code/turf/turf_flock.dm [hunk 4]
```diff
@@ -324,7 +324,7 @@
 	..()
 	if(!istype(F) || !oldloc)
 		return
-	if(F.client && F.client.check_key(KEY_RUN) && !F.floorrunning && F.resources >= 1)
+	if(!F.floorrunning && F.resources >= 1)
```
Comment (FlameArrow57) [inline code/turf/turf_flock.dm:327]: Is turf Entered() called whenever set_loc() is called passing turf or something else?

File: code/turf/turf_flock.dm [hunk 5]
```diff
@@ -356,6 +356,19 @@
 		else if(!isfeathertile(newloc))
 			F.end_floorrunning()
 
+/turf/simulated/wall/auto/feather/Bumped(AM)
+	. = ..()
+	if(istype(AM, /mob/living/critter/flock/drone))
+		var/mob/living/critter/flock/drone/F = AM
+		if(F.floorrunning || (F.can_floorrun && F.resources >= 1))
+			if(!F.client || (F.client && F.client.check_key(KEY_RUN))) //ai doesn't have to hold shift to wallrun, people do
```
Comment (FlameArrow57) [inline code/turf/turf_flock.dm:364]: If bumped can be called in more cases than the player purposefully bumping into wall, feel like is_npc and dormant checks might be safer than a no client check?

File: code/turf/turf_flock.dm (hunk 3 above)
Comment (amylizzle) [reply code/turf/turf_flock.dm:366]: it needs to be in that order because `set_loc` triggers `Entered`, which has bonus floorrunning behaviour

**Comment by amylizzle** [comment]
marked as draft while I fix a minor thing (flock structures aren't pathed through properly)

File: code/turf/turf_flock.dm (hunk 2 above)
Comment (amylizzle) [reply code/turf/turf_flock.dm:363]: `can_floorrun` basically just acts as a flag for whether you're in a grab or not. So this if statement is really: "Are you floorrunning already, or are you not grabbed and have resources to start floorrunning?"

File: code/turf/turf_flock.dm (hunk 4 above)
Comment (amylizzle) [reply code/turf/turf_flock.dm:327]: `set_loc()`

File: code/turf/turf_flock.dm (hunk 5 above)
Comment (amylizzle) [reply code/turf/turf_flock.dm:364]: will fix

File: _std/pathfinding.dm (hunk 1 above)
Comment (amylizzle) [reply _std/pathfinding.dm:419]: will fix

File: code/obj/flock/furniture.dm [hunk 6]
```diff
@@ -344,13 +344,16 @@
...
-/obj/grille/flock/Cross(atom/movable/mover)
+/obj/grille/flock/Crossed(atom/movable/mover)
 	. = ..()
 	var/mob/living/critter/flock/drone/drone = mover
 	if(istype(drone) && !drone.floorrunning)
 		animate_flock_passthrough(mover)
 		. = TRUE
 
+/obj/grille/flock/Cross(atom/movable/mover)
+	. = ..()
```
Comment (TobleroneSwordfish) [inline code/obj/flock/furniture.dm:355]:
```suggestion
```
Since `Cross` shouldn't have side effects calling the parent is pointless

File: _std/pathfinding.dm [hunk 7]
```diff
@@ -446,7 +452,7 @@
 						continue
 					else
 						return FALSE
-				return FALSE
+				return A.Cross(passer)
```
Comment (TobleroneSwordfish) [inline _std/pathfinding.dm:455]: Using return here will stop checking objects if one is found to be crossable, see the weird conditional return below

File: code/obj/flock/structure/flock_structure_parent.dm [hunk 8]
```diff
@@ -261,9 +261,13 @@ ABSTRACT_TYPE(/obj/flock_structure)
...
-/obj/flock_structure/Cross(atom/movable/mover)
+/obj/flock_structure/Crossed(atom/movable/mover)
 	. = ..()
 	var/mob/living/critter/flock/drone/drone = mover
 	if(src.passthrough && istype(drone) && !drone.floorrunning)
 		animate_flock_passthrough(mover)
 		. = TRUE
+
+/obj/flock_structure/Cross(atom/movable/mover)
+	. = ..()
```
Comment (TobleroneSwordfish) [inline code/obj/flock/structure/flock_structure_parent.dm:272]:
```suggestion
```

File: code/obj/window.dm [hunk 9]
```diff
@@ -1064,6 +1064,16 @@
...
 
+/obj/window/auto/feather/Crossed(atom/movable/mover)
+	. = ..()
+	var/mob/living/critter/flock/drone/drone = mover
+	if(istype(drone) && !drone.floorrunning)
+		animate_flock_passthrough(mover)
+		. = TRUE
+
+/obj/window/auto/feather/Cross(atom/movable/mover)
+	. = ..()
```
Comment (TobleroneSwordfish) [inline code/obj/window.dm:1075]:
```suggestion
```

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]
How are you handling drones pathfinding with no resources, won't they get stuck staring at a wall?

**Comment by Sovexe** [comment]
was this profiled?

Also has some bugs
image

**Comment by amylizzle** [comment]
> was this profiled?
> 
> Also seem to ignore windows, intended? image
> 
> invisible in walls sometimes? image

It was profiled, and made only a tiny difference in the distribution of get_path_to timings, but did not noticeably increase the timing. I'll do a proper before/after comparison and post results when I've bugfixed.

Invisible in walls is because the new wall sprites don't have an "on" animation. It's coded, just no sprite. 

The major bug here is that the drones aren't supposed to be able to do stuff when floorrunning. I'm looking at fixes now.

**Comment by amylizzle** [comment]
> How are you handling drones pathfinding with no resources, won't they get stuck staring at a wall?

There's a check for resources, which I'll probably up from 1 to 10 or so, but yes - if the path is very long and goes through multiple walls it will fail to path, and then have to recalculate the path from where it got stuck. Although movement tasks don't currently have a timeout, which they probably should. Will add.

**Comment by amylizzle** [comment]
Profiling:
### BEFORE
one sec 
### AFTER
image
image

**Comment by amylizzle** [comment]
Ran it for a bit longer and got this:
image

tl;dr the error on the profiler is too great to see a measurable effect at these levels.

**Comment by amylizzle** [comment]
Comparison: (external trace is upstream/master, this trace is this pr)
image

File: code/obj/flock/furniture.dm [hunk 10]
```diff
@@ -34,6 +34,18 @@
...
+	var/mob/living/critter/flock/drone/drone = mover
+	if(istype(drone) && !drone.floorrunning)
+		animate_flock_passthrough(mover)
+		. = TRUE
+	else if(istype(mover,/mob/living/critter/flock))
+		. = TRUE
+
+/obj/table/flock/Cross(atom/movable/mover)
+	return istype(mover,/mob/living/critter/flock)
+
```
Comment (FlameArrow57) [inline code/obj/flock/furniture.dm:48]: Do drones really need to pass through tables?

File: code/mob/living/critter/flock/flockdrone.dm [hunk 11]
```diff
@@ -731,6 +737,9 @@
 	if(src.resources < FLOCK_LAY_EGG_COST)
 		boutput(src, "<span class='alert'>Not enough resources (you need [FLOCK_LAY_EGG_COST]).</span>")
 		return
+	if(src.floorrunning)
+		boutput(src, "<span class='alert'>You can't do that while floor running.</span>")
```
Comment (FlameArrow57) [inline code/mob/living/critter/flock/flockdrone.dm:741]:
```suggestion
		boutput(src, "<span class='alert'>You can't do that while floorrunning.</span>")
```
```suggestion
		boutput(src, "<span class='alert'>You can't do that while floor running.</span>")
```

File: code/mob/living/critter/flock/flockdrone.dm [hunk 12]
```diff
@@ -731,6 +737,9 @@
 	if(src.resources < FLOCK_LAY_EGG_COST)
 		boutput(src, "<span class='alert'>Not enough resources (you need [FLOCK_LAY_EGG_COST]).</span>")
 		return
+	if(src.floorrunning)
```
Comment (FlameArrow57) [inline code/mob/living/critter/flock/flockdrone.dm:740]: What about restrained check?

File: code/turf/turf_flock.dm (hunk 4 above)
Comment (FlameArrow57) [reply code/turf/turf_flock.dm:327]: Couldn't there be possible issues, such as wormholes, making you auto-floorrun if you teleport into a wall without a key check?

**Comment by FlameArrow57** [comment]
Idk, kind of against floorrunning through windows. There are two versions of flock windows, unless auto inherits procs from regular, would ideally need to maintain two versions of that code. Also, not sure how it would look making a sprite for on windows. Not sure if it's in theme too, as Flock windows are just gnesis glass, compared to floors and walls which are direct conduits. Also, it's almost guaranteed that there will be a Flock floor that can be floorran underneath a Flock window, I don't really see a need to make floorrunning through windows a thing

File: code/obj/flock/furniture.dm (hunk 10 above)
Comment (amylizzle) [reply code/obj/flock/furniture.dm:48]: it mostly just makes them consistent with other flock things (buildings, barricades, etc). I did it as a result of play testing, and I feel like it improved things.

File: code/mob/living/critter/flock/flockdrone.dm [hunk 13]
```diff
@@ -505,6 +505,12 @@
 			if (wall.on)
 				wall.off()
 	animate_flock_floorrun_end(src)
+	if (istype(src.loc, /turf/simulated/wall/auto/feather))
+		for(var/turf/T in getneighbours(src.loc))
+			if(!flock_is_blocked_turf(T))
+				src.set_loc(T)
+				return
+
```
[the remark this answers is not in the document]
Comment (amylizzle) [reply code/mob/living/critter/flock/flockdrone.dm:513]: yeah, it makes them basically immune to projectile weapons.

File: code/mob/living/critter/flock/flockdrone.dm (hunk 12 above)
Comment (amylizzle) [reply code/mob/living/critter/flock/flockdrone.dm:740]: what do you mean?

**Comment by amylizzle** [comment]
> Idk, kind of against floorrunning through windows. There are two versions of flock windows, unless auto inherits procs from regular, would ideally need to maintain two versions of that code. Also, not sure how it would look making a sprite for on windows. Not sure if it's in theme too, as Flock windows are just gnesis glass, compared to floors and walls which are direct conduits. Also, it's almost guaranteed that there will be a Flock floortile that can be floorran underneath a Flock window, I don't really see a need to make floorrunning through windows a thing

I haven't actually changed how floorrunning through windows works, except to bring the behaviour a little more in line with walls for reasons of pathing. You could floorrun under them before, and you can now. The only actual difference is that now you can start floorrunning automatically by running into a window the same as you can a wall. 

As for the multiple types, they aren't actually used outside of set pieces which is fine. No point adding the behaviour to them, so no need to maintain both versions of code.

File: code/mob/living/critter/flock/flockdrone.dm (hunk 13 above)
[the remark this answers is not in the document]
Comment (FlameArrow57) [reply code/mob/living/critter/flock/flockdrone.dm:513]: Okay, feel like it would be good to let them just stay in walls considering it isn't like that for any other mob, but I guess it's a bit different with it being much easier for Flockdrones. In a way though, you could be just as, or nearly as, tricky with a Flockdrone even if this weren't implemented, since you could go behind and in front of the wall constantly

**Comment by FlameArrow57** [comment]
If it's to make windows a little more similar to walls, what about making it instead so that bumping into a window just makes you auto floorrun into a Flock floortile under the window if there is one?

Also, if window floorrunning is went with, need to add window stuff to at least Flockdrone process_move()

File: code/mob/living/critter/flock/flockdrone.dm (hunk 13 above)
[the remark this answers is not in the document]
Comment (TobleroneSwordfish) [reply code/mob/living/critter/flock/flockdrone.dm:513]: Can they shoot out while in walls? If so then definitely kick them out, we do not want invincible wall drones

[the remark this answers is not in the document]
Comment (FlameArrow57) [reply code/mob/living/critter/flock/flockdrone.dm:513]: Think you could do the same with almost anything else you could Floorrun into though, like Flock structures and soon tables/closets if implemented

**Comment by amylizzle** [comment]
> If it's to make windows a little more similar to walls, what about making it instead so that bumping into a window just makes you auto floorrun into a Flock floortile under the window if there is one?

That is a much less dumb way of doing it. Will fix. Derp.

File: code/obj/flock/furniture.dm (hunk 10 above)
Comment (amylizzle) [reply code/obj/flock/furniture.dm:48]: It shouldn't need to be added, `jpsTurfPassable()` checks for `Cross()` on all objects in a turf, so just adding `Cross() return istype(flock)` is sufficient to make any item pathable through for flock critters.

File: code/mob/living/critter/flock/flockdrone.dm (hunk 12 above)
Comment (amylizzle) [reply code/mob/living/critter/flock/flockdrone.dm:740]: probably better to add that as an extra condition in future, otherwise being handcuffed while trying to lay an egg as a drone would give you a message about floorrunning. Also I can't think why being handcuffed would prevent egg laying.

File: code/obj/window.dm [hunk 14]
```diff
@@ -1046,6 +1046,7 @@
 // flock windows
 
 /obj/window/auto/feather
+	var/on = FALSE
```
Comment (FlameArrow57) [inline code/obj/window.dm:1049]:
```suggestion
```

File: code/obj/window.dm [hunk 15]
```diff
@@ -1064,6 +1065,19 @@
 /obj/window/auto/feather/proc/repair()
 	src.health = min(src.health + 10, src.health_max)
 
+/obj/window/auto/feather/Crossed(atom/movable/mover)
```
Comment (FlameArrow57) [inline code/obj/window.dm:1068]: Crossed() or Bumped()?

File: code/mob/living/critter/flock/flockdrone.dm (hunk 12 above)
Comment (FlameArrow57) [reply code/mob/living/critter/flock/flockdrone.dm:740]: Flockdrones can't actually be handcuffed

**Comment by FlameArrow57** [comment]
What about Crossed() and Cross() for closets too?

File: code/obj/window.dm (hunk 15 above)
Comment (amylizzle) [reply code/obj/window.dm:1067]: Crossed in this case. Not entirely sure why, but if Cross evaluates to true, Bumped is never triggered. I don't need to do that fir the walls because I modified the pathing algo, and they're turfs, but windows are objects on turfs so Cross/ed it is

File: code/obj/flock/furniture.dm [hunk 16]
```diff
@@ -223,6 +223,19 @@
 		<br><span class='bold'>System Integrity:</span> [round((src.health_attack/src.health_max)*100)]%
 		<br><span class='bold'>###=-</span></span>"}
 
+// flockdrones can always move through
+/obj/storage/closet/flock/Crossed(atom/movable/mover)
+	. = ..()
+	var/mob/living/critter/flock/drone/drone = mover
+	if(!src.open && istype(drone) && !drone.floorrunning)
```
Comment (FlameArrow57) [inline code/obj/flock/furniture.dm:230]: Not sure if matters rn, but feel like the open check should be before, otherwise will ignore open for flock critters

Comment (amylizzle) [reply code/obj/flock/furniture.dm:230]: it's just the animation that gets played for drones. I don't want the animation to play if the closet is open, but they should be able to pass through anyway. Flockbits don't get animated by anything else for some reason, so this is for consistency.

Comment (FlameArrow57) [reply code/obj/flock/furniture.dm:230]: Err I guess concern was that for moving while inside a closet, will Flockdrones kick open closet, or pass through? Also, could move open check into the if check then?

Comment (amylizzle) [reply code/obj/flock/furniture.dm:230]: If you're in the closet and move, it'll open the closet. If the closet is closed and you move onto it, it'll play the flock squeeze animation and not open it. If the closet is open, you'll just move onto it as normal.

>Also, could move open check into the if check then?
which check?

File: code/obj/flock/furniture.dm [hunk 17]
```diff
@@ -211,6 +223,19 @@
...
+		animate_flock_passthrough(mover)
```
Comment (FlameArrow57) [inline code/obj/flock/furniture.dm:231]:
```suggestion
	if(istype(drone) && !drone.floorrunning)
		if (!src.open)
			animate_flock_passthrough(mover)
```

Comment (FlameArrow57) [reply code/obj/flock/furniture.dm:231]: Or src.open check could just be placed earlier, as an early return type of thing

Comment (amylizzle) [reply code/obj/flock/furniture.dm:231]: literally the exact same behaviour, just slower. If `src.open` is `TRUE`, then the parent function will evaluate to `TRUE`. This change would just mean we did and extra `istype()` and a boolean check when the closet is already open that we don't need to do.

**Comment by TobleroneSwordfish** [comment]
Okay after some quick testing, I've noticed that drones seem to think they can act while floorrunning, leading to them getting stuck walking into a wall to get an item over and over
image

They also seem to stay floorrunning while stationary for a lot longer than they should, ideally they should pop out as soon as they finish their path (mobai jank may not allow this though, not a huge deal)

File: code/obj/flock/furniture.dm (hunk 17 above)
Comment (amylizzle) [reply code/obj/flock/furniture.dm:231]: also worth noting than in an if statement, byond does optimisations like in `if(false_func() & true_func())` because `false_func()` returns `FALSE`, `true_func()` is never called, meaning `if(!src.open & ...` is equivalent to an early return.

**Review by TobleroneSwordfish — APPROVED** [review]
Issues fixed, profiles fine, lgtm

File: code/mob/living/critter/ai/flock/flocktasks.dm [hunk 18]
```diff
@@ -139,7 +139,7 @@ stare
 	. = list()
 	for(var/turf/simulated/floor/feather/F in view(max_dist, holder.owner))
 		// let's not spam eggs all the time
-		if(isnull(locate(/obj/flock_structure/egg) in F))
+		if(!is_blocked_turf(F) && isnull(locate(/obj/flock_structure/egg) in F))
```
Comment (FlameArrow57) [inline code/mob/living/critter/ai/flock/flocktasks.dm:142]: Is precedence of in > &&?

Comment (amylizzle) [reply code/mob/living/critter/ai/flock/flocktasks.dm:142]: `!is_blocked_turf(F) && isnull()`

Comment (FlameArrow57) [reply code/mob/living/critter/ai/flock/flocktasks.dm:142]: Oh, I misread the parenthesis, nvm
